cmd/upload.go

```diff
@@ -13,30 +13,248 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
-
-// rekor upload --signature=acme.sig --public-key=acme.pub --artifact=https://acmeproject/acme123.tar.gz | ./acme123.tar.gz --sha
 package cmd
 
 import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/google/trillian"
+	homedir "github.com/mitchellh/go-homedir"
+	"github.com/projectrekor/rekor-cli/app"
 	"github.com/projectrekor/rekor-cli/log"
 	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"golang.org/x/crypto/openpgp"
+	"golang.org/x/crypto/openpgp/armor"
 )
 
+type RespStatusCode struct {
+	Code string `json:"file_recieved"`
+}
+
+type getLeafResponse struct {
+	Status RespStatusCode
+	Leaf   *trillian.GetLeavesByIndexResponse
+	Key    []byte
+}
+
+type RekorEntry struct {
+	SHA       string `json:"SHA,omitempty"`
+	URL       string `json:"URL,omitempty"`
+	Signature []byte `json:"Signature"`
+	PublicKey []byte `json:"PublicKey"`
+}
+
+type RekorArmorEntry struct {
```

Comment:
RekorArmorEntry and RekorEntry are the same afaict, why do we need both types?

Reply:
Sure, they use different types for Signature and PublicKey

RekorEntry:
```
	Signature []byte `json:"Signature"`
	PublicKey []byte `json:"PublicKey"`
```
RekorArmorEntry:
```
	Signature string `json:"Signature"`
	PublicKey string `json:"PublicKey"`
```

This is so that we can handle armored and binary gpg artifacts.

Reply:
Ah, what do you think about merging these two into one struct?

```go
type RekorEntry struct {
	SHA       string `json:"SHA,omitempty"`
	URL       string `json:"URL,omitempty"`
	Signature []byte `json:"Signature"`
	PublicKey []byte `json:"PublicKey"`
        AmoredSignature string `json:"Signature"`
	AmoredPublicKey string `json:"PublicKey"`
}
```

Reply:
I guess so, will it not fail if a key is missing? For example there is no value for `AmoredPublicKey` as its a binary instance?